### PR TITLE
Core pipeline; start erasure work

### PR DESCRIPTION
**Pull request by cwgoes**
Ref https://github.com/cryptiumlabs/juvix/issues/112#issuecomment-542190055

This PR does the following:
- Adds morley as a dependency in preparation for later Michelson backend support.
- Parameterises elementary affine core over primitive types & primitive values.
- Adds an evaluator for erased core.
- Adds a translator from erased core to the inet representation (wip, see https://github.com/cryptiumlabs/juvix/issues/158)
- Updates the core HR/IR/erased/EAC pipeline to match the language reference.
- Moves some packages around for clarity (e.g. HR, IR, EAC, Erased Core are all "Core").
- Updates the interpreter (`juvix interactive`).

Erasure is not finished as it requires threading-through the core typechecker in a rather complex manner - I think that will be best as a separate PR. Ref #159 & #160.

### CODE REVIEW

**Comment by cwgoes** [comment]
- [ ] Parameterise EAC over `primVal`, use erased core terms as basis
- [ ] Update EAC constraint solving algorithm accordingly
- [ ] Ensure EAC check outputs erased core
- [ ] Move EAC logic into Core package, as it is a 'flavor'
- [ ] Add erasure testcases
- [ ] Think about how to update BOHM
- [ ] Think about how to interface LLVM

**Comment by cwgoes** [comment]
Ref https://github.com/cryptiumlabs/juvix/issues/159

**Comment by cwgoes** [comment]
To minimise massive PRs I am going to split the rest of this work to https://github.com/cryptiumlabs/juvix/issues/160.

File: src/Juvix/Bohm.hs [hunk 1]
```diff
@@ -3,10 +3,21 @@ module Juvix.Bohm
...
 import Juvix.Bohm.Translation
 import Juvix.Bohm.Type
+import qualified Juvix.Core.Erased.Types as Erased
+import Juvix.Library
+
+erasedCoreToBohm ∷ ∀ primVal. Erased.Term primVal → Bohm
+erasedCoreToBohm term =
+  case term of
+    Erased.Var s → Symbol' s
+    Erased.Prim _ → undefined
```
Comment (mariari) [inline src/Juvix/Bohm.hs]: What does Prim present here? If there is nothing we can change this to safely, let's make the type `Either ??? Bohm`

File: src/Juvix/Core/EAC/Check.hs [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
-import Juvix.EAC.Solve
-import Juvix.EAC.Types
+import Juvix.Core.EAC.ConstraintGen
+import Juvix.Core.EAC.Solve
+import Juvix.Core.EAC.Types
+import Juvix.Core.Erased.Types
 import Juvix.Library hiding (link, reduce)
 
-validEal ∷ Term → TypeAssignment → IO (Either Errors (RPT, ParamTypeAssignment))
+validEal ∷ ∀ primTy primVal. (Eq primTy) ⇒ Term primVal → TypeAssignment primTy → IO (Either (Errors primTy primVal) (RPT primVal, ParamTypeAssignment primTy))
```
Comment (mariari) [inline src/Juvix/Core/EAC/Check.hs]: Should probably make this multiple lines, it gets past 159 characters long

File: src/Juvix/Core/EAC/ConstraintGen.hs [hunk 3]
```diff
@@ -12,7 +13,7 @@ import Prelude (error)
 {- Main functionality. -}
 
 -- Construct occurrence map.
-setOccurrenceMap ∷ (HasState "occurrenceMap" OccurrenceMap m) ⇒ Term → m ()
+setOccurrenceMap ∷ ∀ m primVal. (HasState "occurrenceMap" OccurrenceMap m) ⇒ Term primVal → m ()
```
Comment (mariari) [inline src/Juvix/Core/EAC/ConstraintGen.hs]: How vital are these foralls? are they needed now due to paramaterization?

File: src/Juvix/Core/EAC/ConstraintGen.hs [hunk 4]
```diff
@@ -196,15 +203,16 @@ boxAndTypeConstraint parameterizedAssignment term = do
 
 -- Generate constraints.
 generateTypeAndConstraints ∷
+  ∀ m primTy primVal.
   ( HasState "path" Path m,
     HasState "varPaths" VarPaths m,
     HasState "nextParam" Param m,
-    HasState "typeAssignment" TypeAssignment m,
+    HasState "typeAssignment" (TypeAssignment primTy) m,
```
Comment (mariari) [inline src/Juvix/Core/EAC/ConstraintGen.hs]: Some type aliases might be nice here?

File: src/Juvix/Core/Erased/Evaluator.hs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  )
+where
+
+import Juvix.Core.Erased.Types
+import Juvix.Library hiding (Map, evaluate)
+import Juvix.Utility
+
+evaluate ∷
+  ∀ primVal m.
+  ( HasReader "apply" (primVal → primVal → Maybe primVal) m,
```
Comment (mariari) [inline src/Juvix/Core/Erased/Evaluator.hs]: Nice I'm curious how you got this up, I could't make this from a state monad sadly.

File: src/Juvix/Core/Pipeline.hs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  -- Then invoke Z3 to check elementary-affine-ness.
+  start ← liftIO unixTime
+  result ← liftIO (EAC.validEal erased assignment)
+  end ← liftIO unixTime
+  tell @"log" [LogRanZ3 (end - start)]
+  -- Return accordingly.
+  case result of
+    Right (eac, _) → do
+      let erasedEac = EAC.erase eac
+      unless (erasedEac == erased) (throw @"error" (InternalInconsistencyError "erased affine core should always match erased core"))
```
Comment (mariari) [inline src/Juvix/Core/Pipeline.hs]: Another nitpick about a new line

File: src/Juvix/Core/EAC/ConstraintGen.hs (hunk 3 above)
Comment (cwgoes) [reply src/Juvix/Core/EAC/ConstraintGen.hs]: It was always forall, I just made it explicit.

File: src/Juvix/Core/EAC/ConstraintGen.hs (hunk 4 above)
Comment (cwgoes) [reply src/Juvix/Core/EAC/ConstraintGen.hs]: How so? They would need to be scoped to the function.

File: src/Juvix/Core/Erased/Evaluator.hs (hunk 5 above)
Comment (cwgoes) [reply src/Juvix/Core/Erased/Evaluator.hs]: Alas, I just haven't made a monad which actually uses this yet.

File: src/Juvix/Core/EAC/Check.hs (hunk 2 above)
Comment (cwgoes) [reply src/Juvix/Core/EAC/Check.hs]: Yup, updated.

File: src/Juvix/Core/Pipeline.hs (hunk 6 above)
Comment (cwgoes) [reply src/Juvix/Core/Pipeline.hs]: Added a newline.

File: src/Juvix/Bohm.hs (hunk 1 above)
Comment (cwgoes) [reply src/Juvix/Bohm.hs]: In the follow-up PR (#158) I will fix this so that it is complete.

**Pull request merged**